[lldb][rpc] Upstream RPC Client Library Emitters

chelcassanova opened the pull request:
This commit upstreams the client side emitters for the lldb-rpc-gen tool alongside tests for its functionality and heuristics.

https://discourse.llvm.org/t/rfc-upstreaming-lldb-rpc/85804

DavidSpickett commented on lldb/test/Shell/RPC/Generator/Inputs/Client/CheckArrayPointer.h:
```diff
@@ -0,0 +1,20 @@
+#include "lldb/API/SBDefines.h"
+
+namespace lldb {
+class LLDB_API SBRPC_CHECKARRAYPTR {
```
So these are just test inputs, and they each have a method that should be handled in some specific way. I presume the RPC tool works on classes, so we have to wrap it in some class at least.

I'd prefer the standard class naming, even though it's just test code. SBRPC_CheckArrayPtr is fine.

ALLCAPS makes me think it's acting as if it were a define.

medismailben replied:
+1

DavidSpickett commented on lldb/test/Shell/RPC/Generator/Inputs/Client/CheckArrayPointer.h:
```diff
@@ -0,0 +1,20 @@
+#define LLDB_API_SBRPC_CHECKARRAYPTR_H
+
+#include <cstddef>
+#include <cstdio>
```
Oh these are just test inputs. So the include can be removed entirely?

medismailben replied:
Make sure to remove this from every test

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/client/RPCLibrarySourceEmitter.cpp:
```diff
@@ -0,0 +1,542 @@
+      EmitLine(
+          "RPCValueEncoder(send, rpc_common::RPCPacket::ValueType::Argument, " +
+          StringListName + ");");
+      // When we have pointer parameters, in general the strategy is
+      // to create `Bytes` objects from them (basically a buffer with a size)
+      // and then move those over the wire. We're not moving the pointer itself,
+      // but the contents of memory being pointed to. There are a few exceptions
+      // to this:
+      //   - If the type is `const char *` or `const char **`, those are handled
+      //     specially and can be encoded directly.
+      //   - If we have a function pointer, we move the pointer value directly.
+      //     To do the callback from the server-side, we will need this pointer
+      //     value to correctly invoke the function client-side.
+      //   - If we have a baton (to support a callback), we need to move the
+      //     pointer value directly. This is for the same reason as callbacks
+      //     above.
+      //   - If we have a pointer to an SB class, we just dereference it and
+      //     encode it like a normal SB class object.
```
Or fix the tab spaces mix to make it match visually, it could be that.

chelcassanova replied:
I'm actually going to move this comment to be inside of the else if block below and match it with the comment that's already in that block.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/client/RPCLibrarySourceEmitter.cpp:
```diff
@@ -0,0 +1,542 @@
+    EmitLine("RPCValueEncoder(send, "
+             "rpc_common::RPCPacket::ValueType::Argument, *this);");
+
+  for (auto Iter = method.Params.begin(); Iter != method.Params.end(); Iter++) {
```
If Iter is not modified within the loop, use a range based for if you can make it work.

DavidSpickett replied:
Oh no it is modified.
```
      Iter++;
      StringListLine += Iter->Name + ");";
      Iter--;
```
Which is some, shall I say, "interesting" code :)

In this case you could use std::next(Iter)->Name rather than undoing the increment.

And you can't use a range based for because you need to peek forwards, but certainly you can clean up the handling of Iter a bit.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/client/RPCLibrarySourceEmitter.cpp:
```diff
@@ -0,0 +1,542 @@
+      IndentLevel--;
+      EmitLine("else");
+      IndentLevel++;
+      // FIXME: change this logic, not everything can be an rpc::ObjectRef!!!
```
Do we intend to do something about this?

chelcassanova commented on lldb/tools/lldb-rpc/lldb-rpc-gen/client/RPCLibraryHeaderEmitter.cpp:
```diff
@@ -0,0 +1,124 @@
+    std::string EnumValue = EC->getNameAsString();
+    SmallString<16> ValueStr;
+    EC->getInitVal().toString(ValueStr);
+    EnumValue += " = " + ValueStr.str().str() + ", ";
```
`SmallString` actually has a `std::string` operator so I think can get replaced with `std::string(ValueStr)` here.

chelcassanova committed:
[lldb][rpc] Add RPC client library tests

Adds shell tests for the LLDB RPC client library.

chelcassanova committed:
[lldb][rpc] Add RPC client library emitters

Adds the RPC client library sources and header emitters.

chelcassanova committed:
[lldb][rpc] Update lldb-rpc-gen tool to account for client

Updates the lldb-rpc-gen tool to account for everything necessary to
emit the client library code.

medismailben commented on lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:
```diff
@@ -46,6 +48,12 @@ static std::string GetServerOutputDirectory() {
 }
 
+static std::string GetLibraryOutputDirectory() {
+  llvm::SmallString<128> Path(OutputDir.getValue());
```
nit: I was wondering if we should make the template argument here be `PATH_MAX` but that's only defined on unix platforms, on windows is another macro called `MAX_PATH` which is hard coded to 260 I believe. May be we should go with that to match the smallest path length.

medismailben commented on lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:
```diff
@@ -306,6 +348,28 @@ bool EmitAmalgamatedServerHeader(const std::vector<std::string> &Files) {
   return true;
 }
 
+bool EmitAmalgamatedLibraryHeader(const std::vector<std::string> &Files) {
```
Not sure what "Amalgamated" means here, I'd like a more self-explaining name or a least a comment.

chelcassanova replied:
For RPC, the amalgamated library header is LLDBRPC.h, similar to LLDB.h. I can add a comment here.

medismailben commented:
Looks like there is a linking issue on Linux:

```
2025-07-24T23:47:48.0834355Z [6326/6332] Linking CXX executable bin/lldb-rpc-gen
2025-07-24T23:47:48.0834814Z FAILED: bin/lldb-rpc-gen 
2025-07-24T23:47:48.0852723Z : && /opt/llvm/bin/clang++ -gmlt -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-unknown-pragmas -Wno-strict-aliasing -Wno-vla-extension -O3 -DNDEBUG -fuse-ld=lld -Wl,--color-diagnostics    -Wl,--gc-sections tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/RPCCommon.cpp.o tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/server/RPCServerHeaderEmitter.cpp.o tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/server/RPCServerSourceEmitter.cpp.o tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o -o bin/lldb-rpc-gen  -Wl,-rpath,"\$ORIGIN/../lib:"  lib/libLLVMSupport.a  lib/libclangAST.a  lib/libclangBasic.a  lib/libclangCodeGen.a  lib/libclangFrontend.a  lib/libclangLex.a  lib/libclangRewrite.a  lib/libclangSerialization.a  lib/libclangTooling.a  lib/libLLVMCoverage.a  lib/libLLVMFrontendDriver.a  lib/libLLVMLTO.a  lib/libLLVMExtensions.a  lib/libLLVMPasses.a  lib/libLLVMCoroutines.a  lib/libLLVMHipStdPar.a  lib/libLLVMipo.a  lib/libLLVMLinker.a  lib/libLLVMIRPrinter.a  lib/libLLVMInstrumentation.a  lib/libLLVMVectorize.a  lib/libLLVMSandboxIR.a  lib/libLLVMCFGuard.a  lib/libLLVMGlobalISel.a  lib/libLLVMSelectionDAG.a  lib/libLLVMCodeGen.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMObjCARCOpts.a  lib/libLLVMCGData.a  lib/libLLVMBitWriter.a  lib/libLLVMTarget.a  lib/libclangFrontend.a  lib/libclangParse.a  lib/libclangSerialization.a  lib/libclangSema.a  lib/libclangAnalysis.a  lib/libclangAPINotes.a  lib/libclangEdit.a  lib/libclangSupport.a  lib/libclangDriver.a  lib/libLLVMWindowsDriver.a  lib/libLLVMOption.a  lib/libclangASTMatchers.a  lib/libclangAST.a  lib/libLLVMFrontendHLSL.a  lib/libclangFormat.a  lib/libclangToolingInclusions.a  lib/libclangToolingCore.a  lib/libclangRewrite.a  lib/libclangLex.a  lib/libclangBasic.a  lib/libLLVMFrontendOpenMP.a  lib/libLLVMFrontendOffloading.a  lib/libLLVMObjectYAML.a  lib/libLLVMScalarOpts.a  lib/libLLVMAggressiveInstCombine.a  lib/libLLVMInstCombine.a  lib/libLLVMTransformUtils.a  lib/libLLVMFrontendAtomic.a  lib/libLLVMAnalysis.a  lib/libLLVMProfileData.a  lib/libLLVMSymbolize.a  lib/libLLVMDebugInfoGSYM.a  lib/libLLVMDebugInfoPDB.a  lib/libLLVMDebugInfoCodeView.a  lib/libLLVMDebugInfoMSF.a  lib/libLLVMDebugInfoBTF.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMObject.a  lib/libLLVMIRReader.a  lib/libLLVMBitReader.a  lib/libLLVMAsmParser.a  lib/libLLVMCore.a  lib/libLLVMRemarks.a  lib/libLLVMBitstreamReader.a  lib/libLLVMMCParser.a  lib/libLLVMMC.a  lib/libLLVMTextAPI.a  lib/libLLVMDebugInfoDWARFLowLevel.a  lib/libLLVMBinaryFormat.a  lib/libLLVMFrontendDirective.a  lib/libLLVMTargetParser.a  lib/libLLVMSupport.a  -lrt  -ldl  -lm  /usr/lib/x86_64-linux-gnu/libz.so  lib/libLLVMDemangle.a && :
2025-07-24T23:47:48.0865497Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibraryHeaderEmitter::Begin()
2025-07-24T23:47:48.0866311Z >>> referenced by RPCLibraryHeaderEmitter.h:17 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/client/RPCLibraryHeaderEmitter.h:17)
2025-07-24T23:47:48.0867384Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBAction::CreateASTConsumer(clang::CompilerInstance&, llvm::StringRef))
2025-07-24T23:47:48.0867932Z 
2025-07-24T23:47:48.0870989Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibraryHeaderEmitter::End()
2025-07-24T23:47:48.0872025Z >>> referenced by RPCLibraryHeaderEmitter.h:20 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/client/RPCLibraryHeaderEmitter.h:20)
2025-07-24T23:47:48.0872953Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::~SBVisitor())
2025-07-24T23:47:48.0873420Z 
2025-07-24T23:47:48.0874065Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibraryHeaderEmitter::StartClass(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>)
2025-07-24T23:47:48.0874978Z >>> referenced by lldb-rpc-gen.cpp:111 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:111)
2025-07-24T23:47:48.0875858Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0876318Z 
2025-07-24T23:47:48.0876746Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibrarySourceEmitter::StartClass(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>)
2025-07-24T23:47:48.0877625Z >>> referenced by lldb-rpc-gen.cpp:112 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:112)
2025-07-24T23:47:48.0878469Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0878921Z 
2025-07-24T23:47:48.0879166Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibraryHeaderEmitter::EmitEnum(clang::EnumDecl*)
2025-07-24T23:47:48.0880029Z >>> referenced by lldb-rpc-gen.cpp:115 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:115)
2025-07-24T23:47:48.0880851Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0881306Z 
2025-07-24T23:47:48.0881521Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibraryHeaderEmitter::EndClass()
2025-07-24T23:47:48.0882169Z >>> referenced by lldb-rpc-gen.cpp:133 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:133)
2025-07-24T23:47:48.0883004Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0883516Z 
2025-07-24T23:47:48.0883901Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibrarySourceEmitter::EndClass()
2025-07-24T23:47:48.0884564Z >>> referenced by lldb-rpc-gen.cpp:134 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:134)
2025-07-24T23:47:48.0885399Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0885851Z 
2025-07-24T23:47:48.0886140Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibrarySourceEmitter::EmitMethod(lldb_rpc_gen::Method const&)
2025-07-24T23:47:48.0886873Z >>> referenced by lldb-rpc-gen.cpp:127 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:127)
2025-07-24T23:47:48.0887705Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0888151Z 
2025-07-24T23:47:48.0888441Z ld.lld: error: undefined symbol: lldb_rpc_gen::RPCLibraryHeaderEmitter::EmitMethod(lldb_rpc_gen::Method const&)
2025-07-24T23:47:48.0889174Z >>> referenced by lldb-rpc-gen.cpp:128 (/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/tools/lldb-rpc-gen/lldb-rpc-gen.cpp:128)
2025-07-24T23:47:48.0890010Z >>>               tools/lldb/tools/lldb-rpc-gen/CMakeFiles/lldb-rpc-gen.dir/lldb-rpc-gen.cpp.o:(SBVisitor::VisitCXXRecordDecl(clang::CXXRecordDecl*))
2025-07-24T23:47:48.0890655Z clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

medismailben reviewed (approved):
Overall, LGTM with comments and linking failure addressed.

chelcassanova committed:
[lldb][rpc] Update RPC CMake to add client emitters

Updates the CMake files for RPC to account for the client emitters

DavidSpickett commented on lldb/test/Shell/RPC/Generator/Tests/Client/CheckSBPointer.test:
```diff
@@ -0,0 +1,15 @@
+# class must be created and encoded. Otherwise, the original
+# parameter will be encoded.
+CHECK: lldb_rpc::SBRPC_CHECKSBPTR::CheckSBPtr
+CHECK: if (debugger_ptr)
```
I assume the if / else can use CHECK-NEXT?

DavidSpickett reviewed (commented):
Please go over the outstanding comments and make sure they have been done. I see several that have not.

DavidSpickett reviewed (approved):
LGTM
